test/python/pulse/test_schedule.py

```diff
@@ -946,18 +958,21 @@ def test_single_channel_out_of_order(self):
                         (15, Play(Waveform(np.ones(10)), DriveChannel(0))),
                         (5, Play(Waveform(np.ones(10)), DriveChannel(0)))]
 
+        # ToDo(4872) remove the use of deprecated constructor Schedule(*schedules)
         self.assertEqual(Schedule(*instructions), Schedule(*reversed(instructions)))
 
     def test_multiple_channels_out_of_order(self):
         """Test that schedule with multiple channels equal when out of order."""
         instructions = [(0, ShiftPhase(0, DriveChannel(1))),
                         (1, Acquire(10, AcquireChannel(0), MemorySlot(1)))]
 
+        # ToDo(4872) remove the use of deprecated constructor Schedule(*schedules)
         self.assertEqual(Schedule(*instructions), Schedule(*reversed(instructions)))
 
     def test_different_name_equal(self):
         """Test that names are ignored when checking equality."""
 
+        # ToDo(4872) remove the use of deprecated constructor Schedule(*schedules)
```

Comment:
These changes will certainly make code like this more verbose 😦 . After working through these changes what do you think about them as a whole @mlarocca?

Reply:
Overall I think there were a dozen places where this had to be changed (tests excluded), in various ways.
It's a bit more verbose, of course, but not that bad.
The one thing that worries me a bit is that right now refactoring the code to remove the deprecated constructor is a bit tricky, because if you forget to use `inplace=True` named argument on `.insert`, you mess up - if you have tests then you are golden, you'll notice, but client code might not be as well covered... you know.

So maybe it makes sense to add the `insert_iter` method, as you had suggested?

Reply:
The other option (which may be frustrating :( ) is to leave the current functionality. Why, I had initially proposed removing the instructions int the constructors was to turn `Schedule` into a representation of a function in the pulse IR. I had imagined.
```python
sched = Schedule(signature)
sched.append(Instruction(100, sched.params[0])) 
...
caller = Schedule()
caller.call(sched, args=[1.0])
```
In this way having Schedule be a representation of a function within the IR. However, the way parameters work in circuits and now schedules with #5073 is that the parameters of the object are automatically extended when a new instruction with a parameter is added. This means that having a signature in the constructor would be pointless in these objects as it would be wrong as soon as a new instruction with a parameter would be added.

The workaround to this is to view the `QuantumCircuit`, `Schedule` as basic blocks, that just contain instructions. A `Function` (which doesn't exist currently in Qiskit) should have a signature well defined and be composed of several basic blocks.

The reason I bring this up is that if we are not going to turn `Schedule` into a `Function` with a signature than we don't necessarily have to change the current constructor behaviour. Do you have any thoughts about this?

Reply:
@taalexander 
Thanks again for your feedback! No worries, no frustration - even if we decide to discard these changes, it will have been useful for me to get to know better the code base and how this Schedule/Instruction concept works :)

About your question: my take is that objects for which you pass a signature should indeed be immutable (or otherwise invalidated on `.append`), ideally I would pass both a signature and the instructions/parameters list in the constructor.

So, I agree with you on this, #5073 makes having a signature in the constructor pointless.
I like your idea of going for `Function` and `Schedule` classes. I guess the future `Function` class would be similar to what's currently in schedule.py, did I get that right?

Anyway, bottom line: no problem if you'd like to close this PR and gloss over the changes, I can create a new PR with just the couple of bug fixes and tests that I had added.

How would you like to proceed?
Thanks!